### PR TITLE
move long-prompt weighting code to utils

### PR DESCRIPTION
**Pull request by noskill**
Long-prompt weighting pipeline can't be used with other pipelines e.g. StableDiffusionKDiffusionPipeline

this PR moves long-prompt weighting code to utils so that long-prompt weighting can be used with any pipeline:

```
from diffusers import StableDiffusionKDiffusionPipeline
from diffusers import utils
import types
import torch

pipe = StableDiffusionKDiffusionPipeline.from_pretrained(
      "frankjoshua/icbinpICantBelieveIts_v8",
       torch_dtype=torch.float16
)
pipe.set_scheduler('sample_dpmpp_2m')
pipe=pipe.to("cuda")
# replace encode_prompt in the pipe with long-prompt weighting
pipe.encode_prompt = types.MethodType(utils.encode_prompt, pipe)
```

### CODE REVIEW

**Comment by noskill** [comment]
Also there is code by @takuma104 https://gist.github.com/takuma104/43552b8ec70b63323c57dc9c6fcb9b90, perhaps there should be a community "utils", some code doesn't need it's own pipeline..

**Comment by adhikjoshi** [comment]
this removes the dependency on compel,

File: src/diffusers/utils/lpw.py [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+
+
+import re
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import torch
+from packaging import version
+
+from diffusers.loaders import TextualInversionLoaderMixin, LoraLoaderMixin
+from diffusers.utils import logging
```
Comment (patrickvonplaten) [inline src/diffusers/utils/lpw.py:15]:
```suggestion
from ...loaders import TextualInversionLoaderMixin, LoraLoaderMixin
from .. import logging
```

**Review by patrickvonplaten — COMMENTED** [review]
I'm actually ok with this - wdyt @sayakpaul . Long prompt weighting in A1111 is the main syntax that is being used so it might not be a bad idea to support it in pypi

**Comment by sayakpaul** [comment]
I am okay with that. But not super sure on the maintenance part. I think it should be a bit community-driven to start with. WDYT?

File: src/diffusers/utils/lpw.py (hunk 1 above)
Comment (noskill) [reply src/diffusers/utils/lpw.py:15]: I got an error:
    from ...loaders import TextualInversionLoaderMixin, LoraLoaderMixin
ImportError: attempted relative import beyond top-level package

Comment (noskill) [reply src/diffusers/utils/lpw.py:15]: another error for logging:
```
from .. import logging
ImportError: cannot import name 'logging' from partially initialized module 'diffusers' (most likely due to a circular import) (/home/imgen/miniconda3/envs/py31/lib/python3.10/site-packages/diffusers/__init__.py)
```

**Comment by bghira** [comment]
it seems like a very generic name for something that might only apply to Stable Diffusion pipelines. could this work for, say DeepFloyd or Kandinsky?

File: src/diffusers/utils/lpw.py [hunk 2]
```diff
@@ -0,0 +1,508 @@
+"""
+Implements encode_prompt with long-prompt weighting
+
+adoped from lpw_stable_diffusion by github.com/SkyTNT
```
Comment (bghira) [inline src/diffusers/utils/lpw.py:4]: adopted

**Comment by noskill** [comment]
@bghira generated image changes a little depending on weights, but much less than i see with stable diffusion models. I'll test it with DeepFloyd

```
from diffusers import DiffusionPipeline
from diffusers import utils

import types
import torch


generator = torch.Generator("cuda").manual_seed(3842793274)

pipe_prior = DiffusionPipeline.from_pretrained("kandinsky-community/kandinsky-2-1-prior", torch_dtype=torch.float16)
pipe_prior.to("cuda")

pipe_prior.encode_prompt = types.MethodType(utils.encode_prompt, pipe_prior)

t2i_pipe = DiffusionPipeline.from_pretrained("kandinsky-community/kandinsky-2-1", torch_dtype=torch.float16)
t2i_pipe.to("cuda")
t2i_pipe.encode_prompt = types.MethodType(utils.encode_prompt, t2i_pipe)


prompt = "A alien cheeseburger creature eating itself, claymation, (cinematic:0.7), (moody lighting:0.7)"
negative_prompt = "low quality, bad quality"

image_embeds, negative_image_embeds = pipe_prior(prompt, guidance_scale=6.0, generator=generator).to_tuple()
image = t2i_pipe(
    prompt, image_embeds=image_embeds, negative_image_embeds=negative_image_embeds, height=768, width=768, generator=generator).images[0]
image.save("cheeseburger_monster0.png")
```

**Comment by bghira** [comment]
different versions of Kandinsky have different text encoders:

* 2.2 uses OpenCLIP bigG, same as SDXL refiner and one of SDXL base TEs,
* 2.1 used XLM-Roberta-Large-Vit-L-14, 
* and 2.0 used two small text encoders

i can imagine each behaving differently or breaking with this.

**Comment by noskill** [comment]
Kandinsky 2-2 works with lpw similar to 2-1, image changes a little bit

```
from diffusers import DiffusionPipeline
from diffusers import utils
import functools

import types
import torch


generator = torch.Generator("cuda").manual_seed(3842793274)

pipe_prior = DiffusionPipeline.from_pretrained("kandinsky-community/kandinsky-2-2-prior", torch_dtype=torch.float16)
pipe_prior.to("cuda")

pipe_prior.encode_prompt = types.MethodType(functools.partial(utils.encode_prompt, max_embeddings_multiples=10), pipe_prior)

t2i_pipe = DiffusionPipeline.from_pretrained("kandinsky-community/kandinsky-2-2-decoder", torch_dtype=torch.float16)
t2i_pipe.to("cuda")
t2i_pipe.encode_prompt = types.MethodType(functools.partial(utils.encode_prompt, max_embeddings_multiples=10), t2i_pipe)


prompt = "A alien cheeseburger creature eating itself, claymation, (cinematic:2.7), (moody lighting:2.7)"
negative_prompt = "low quality, bad quality"

image_embeds, negative_image_embeds = pipe_prior(prompt, guidance_scale=6.0, generator=generator).to_tuple()
image = t2i_pipe(image_embeds=image_embeds, negative_image_embeds=negative_image_embeds, height=768, width=768, generator=generator).images[0]
image.save("cheeseburger_monster5.png")
```

**Comment by yijinsheng** [comment]
the sd-xl pipline need prompt_embed_pool,does this support it

**Comment by noskill** [comment]
@yijinsheng no

**Comment by noskill** [comment]
@bghira it is possible to pass embeddings from encode_prompt function to deepfloyd pipeline, but picture quality got much worse:






So from tests it looks like only SD pipelines can benefit from this code so i moved lpw to lpw_stable_diffusion.py

i'll take a look if it is possible to extract common code from stable diffusion and stable diffusion xl lpw pipelines

**Comment by bghira** [comment]
almost undoubtedly it CAN work on the other models, but likely will require investigation into how to do it.

**Comment by patrickvonplaten** [comment]
@yiyixuxu @sayakpaul anybody has time to take over this PR?

**Comment by sarmientoj24** [comment]
@bghira does LPW support kohya style lora prompts `<lora:yyyyy:1.2>`?

**Comment by sayakpaul** [comment]
Prompt weighting is supported via `compel`:
https://huggingface.co/docs/diffusers/using-diffusers/weighted_prompts

**Comment by patrickvonplaten** [comment]
> Prompt weighting is supported via `compel`: https://huggingface.co/docs/diffusers/using-diffusers/weighted_prompts

We can nevertheless support lpw more natively in the diffusers core codebase

**Comment by noskill** [comment]
@sarmientoj24 it doesn't support lora prompts. I would rather add this and other features in a separate PR.

**Comment by adhikjoshi** [comment]
> @sarmientoj24 it doesn't support lora prompts. I would rather add this and other features in a separate PR.


def process_lora_prompt(prompt):
    lora = re.compile(r'<lora:([^:]+):([\d\.\-]+)>')
    lora_matches = lora.findall(prompt)
    lora_matches = [(f'({name}:{weight})', float(weight))
                    for name, weight in lora_matches]
    filtered_prompt = lora.sub('', prompt)
    filtered_prompt += ' '.join(f'({name}:{weight})' for name,
                                weight in lora_matches)
    return filtered_prompt


here is function to extract lora from prompt

**Comment by bghira** [comment]
i don't think that's how you handle LoRA prompting, that merely removes the loras from the prompt.

**Comment by noskill** [comment]
@DN6 @yijinsheng do i need to do anything in this PR? Adding lora and hypernetworks as suggested above is a nice feature but this can be done in follow up PRs..

**Comment by patrickvonplaten** [comment]
@DN6 @yiyixuxu can you give this a look?

**Comment by sayakpaul** [comment]
Not stale.

**Comment by Scorpinaus** [comment]
Should not be stale?

**Comment by sayakpaul** [comment]
@yiyixuxu a gentle bump here.

**Comment by noskill** [comment]
@sayakpaul Maybe we could introduce a PromptEncoder class to pass it into pipeline constructor, or maybe a mixin class?

Using types.MethodType is not the best possible way to change prompt encoding.. but i believe this PR is a step in the right direction

**Comment by sayakpaul** [comment]
Requesting some inputs from @yiyixuxu here.

**Comment by DN6** [comment]
Hi @noskill How did you envision using this in the Pipelines? By just replacing the call to `encode_prompt` with the lpw version of `encode_prompt`?

**Comment by DN6** [comment]
Nvm, I saw missed you snippet using Methods type. PromptProcessor would work I think. It would precompute the embeddings and just pass them directly to the Pipeline. Wdyt @yiyixuxu

**Comment by noskill** [comment]
@DN6 i refactored a bit, now the example looks less hacky

```
from diffusers import StableDiffusionKDiffusionPipeline
from diffusers import utils
import torch

pipe = StableDiffusionKDiffusionPipeline.from_pretrained(
              "frankjoshua/icbinpICantBelieveIts_v8",
                     torch_dtype=torch.float16
                     )
pipe.set_scheduler('sample_dpmpp_2m')
pipe=pipe.to("cuda")
# replace default encode prompt in the pipeline with long-prompt weighting
pipe.set_prompt_encoder(utils.encode_prompt_lpw_sd)
result = pipe('a horse, autumn, (snow)')
result.images[0].save('horse.png')
```

**Comment by DN6** [comment]
Hi @noskill I was thinking something more along the lines of 

```python
from diffusers import StableDiffusionKDiffusionPipeline
from diffusers.utils.prompt_utils import StableDiffusionLongPromptProcessor
import torch

pipe = StableDiffusionKDiffusionPipeline.from_pretrained("frankjoshua/icbinpICantBelieveIts_v8", torch_dtype=torch.float16)
pipe.set_scheduler('sample_dpmpp_2m')
pipe=pipe.to("cuda")

prompt_embeds = StableDiffusionLongPromptProcessor().encode_prompt('a horse, autumn, (snow)')
result = pipe(prompt_embeds=prompt_embeds)
result.images[0].save('horse.png')
```

So we don't have to change the compatible piplines for LPW to work?

**Comment by noskill** [comment]
encode_prompt does a lot of thing, like scaling/unscaling lora layers, so actual code would look more like this:

```
from diffusers import StableDiffusionKDiffusionPipeline
from diffusers.utils.prompt_utils import StableDiffusionLongPromptProcessor
import torch

pipe = StableDiffusionKDiffusionPipeline.from_pretrained("frankjoshua/icbinpICantBelieveIts_v8", torch_dtype=torch.float16)
pipe.set_scheduler('sample_dpmpp_2m')
pipe=pipe.to("cuda")

with StableDiffusionLongPromptProcessor(pipe) as lpw:
        prompt_embeds = lpw.encode_prompt('a horse, autumn, (snow)')
result = pipe(prompt_embeds=prompt_embeds)
result.images[0].save('horse.png')
```

I have code that expects a pipeline to do it's processing so some additional wrapper pipeline is needed with this approach.

**Comment by noskill** [comment]
@DN6  Another idea: maybe we just move lpw code into encode prompt in StableDiffusionXLPipeline and keep it off by default to keep backward compatibility?

**Comment by DN6** [comment]
@noskill This snippet you have provided with the suggested design looks okay to me actually. cc: @yiyixuxu wdyt?  

> I have code that expects a pipeline to do it's processing so some additional wrapper pipeline is needed with this approach.

Sorry not following here? The prompt embeds are being passed to a pipeline right? What additional stuff needs to be done?

**Comment by yiyixuxu** [comment]
agree with @DN6 we should precompute the embeddings and pass it to the pipeline as inputs!

**Comment by yiyixuxu** [comment]
I like the API @DN6 here https://github.com/huggingface/diffusers/pull/5100#issuecomment-2059537449

we probably need to pass the text encoders to it, so maybe this - this is also similar to compel API
```python
lwp_proc = StableDiffusionLongPromptProcessor(pipe.tokenizer, pipe.text_encoder)

prompt_embeds = lpw_proc.encode_prompt('a horse, autumn, (snow)')
result = pipe(prompt_embeds=prompt_embeds)
result.images[0].save('horse.png')
```

**Comment by noskill** [comment]
I refactored code  in src/diffusers/pipelines/stable_diffusion_xl/ by introducing base class. I placed lpw and no-lpw encoding into the base class. From library users perspective it more convenient

```
from diffusers import StableDiffusionXLPipeline

import torch

pipeline = StableDiffusionXLPipeline.from_pretrained("frankjoshua/juggernautXL_v8Rundiffusion", dtype=torch.float16).to('cuda')

prompt = "close up Portrait photo of muscular bearded guy in a worn mech suit, (snow:2.3) , ((light bokeh)), intricate, (steel metal [rust]), elegant, sharp focus, photo by greg rutkowski, soft lighting, vibrant colors, masterpiece, (streets:1.5), detailed face"
negative_prompt = "realistic, nude, cross eyed, tongue, open mouth, inside, 3d, cartoon, anime, sketches, (worst quality:2), (low quality:2), (normal quality:2), lowres, normal quality, ((monochrome)), ((grayscale)), skin spots, acnes, skin blemishes, bad anatomy, red eyes, muscular"

generator = torch.Generator("cuda").manual_seed(74983284384741)

image = pipeline(prompt, width=1024, height=1280, negative_prompt=negative_prompt, num_inference_steps=35, use_lpw=0, generator=generator).images[0]
image.save('no_lpw.png')
generator = torch.Generator("cuda").manual_seed(74983284384741)
image = pipeline(prompt, width=1024, height=1280, negative_prompt=negative_prompt, num_inference_steps=35, use_lpw=1, generator=generator).images[0]
image.save('lpw.png')
```

**Comment by yiyixuxu** [comment]
@noskill 
can you provide one example?  

> encode_prompt(both from lpw and default one) does isinstance checks on the pipeline, it sets some attributes etc.

**Comment by noskill** [comment]
Yes, sure
https://github.com/huggingface/diffusers/blob/ebc99a77aad647c5d33eb36a33c23f7b3949cb40/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L360C17-L360C66

**Comment by noskill** [comment]
@DN6 @sayakpaul Do you guys have any further suggestions where to move with this PR?

**Comment by noskill** [comment]
i'll just reuse get_weighted_text_embeddings_sdxl from lpw_stable_diffusion_xl.py

**Pull request closed without merging**